clinica/iotools/converters/genfi_to_bids/genfi_to_bids_utils.py

```diff
@@ -0,0 +1,586 @@
+        )
+    )
+
+    df_dicom = df_dicom.reset_index()
+
+    df_dicom = df_dicom.assign(source_ses_id=lambda df: df.source_ses_id.astype("int"))
+    df_dicom = df_dicom.assign(
+        genfi_version=lambda df: df.source_ses_id.apply(lambda x: f"GENFI{len(str(x))}")
+    )
+    df_1 = (
+        df_dicom[["source_id", "source_ses_id", "acq_date"]]
+        .groupby(["source_id", "source_ses_id"])
+        .min()
+    )
+    df_2 = df_dicom[["source_id", "acq_date"]].groupby(["source_id"]).min()
+    df_1 = df_1.join(df_2.rename(columns={"acq_date": "baseline"}))
+    df_1 = df_1.assign(
+        ses_month=lambda df: (
+            (
+                df["acq_date"].str[:4].astype("int")
+                - df["baseline"].str[:4].astype("int")
+            )
+            * 12
+            + (
+                df["acq_date"].str[4:6].astype("int")
+                - df["baseline"].str[4:6].astype("int")
+            )
+        )
+    )
+    df_1 = df_1.assign(
+        session_id=lambda df: df.ses_month.map(lambda x: f"ses-M{x:03d}")
+    )
+
+    df_sub_ses = df_dicom.merge(
+        df_1[["baseline", "session_id"]], how="inner", on=["source_id", "source_ses_id"]
+    )
+    df_sub_ses = df_sub_ses.assign(
+        participant_id=lambda df: df.source_id.apply(lambda x: f"sub-{x}")
+    )
+
+    df_sub_ses = df_sub_ses.assign(
+        modality=lambda df: df.series_desc.apply(lambda x: identify_modality(x))
+    )
+    df_sub_ses = df_sub_ses.join(
+        df_sub_ses.modality.map(
+            {
+                "T2w": {
+                    "datatype": "anat",
+                    "suffix": "T2w",
+                    "sidecars": ["T2w.json"],
+                    "task": "",
+                },
+                "T1": {
+                    "datatype": "anat",
+                    "suffix": "T1w",
+                    "sidecars": ["T1w.json"],
+                    "task": "",
+                },
+                "dwi": {
+                    "datatype": "dwi",
+                    "suffix": "dwi",
+                    "sidecars": [
+                        "dwi.json",
+                        "dwi.bval",
+                        "dwi.bvec",
+                    ],
+                    "task": "",
+                },
+                "rsfmri": {
+                    "datatype": "func",
+                    "suffix": "bold",
+                    "sidecars": ["rsfmri.json"],
+                    "task": "_task-rest",
+                },
+                "fieldmap": {
+                    "datatype": "fmap",
+                    "suffix": "fmap",
+                    "sidecars": ["fmap.json"],
+                    "task": "",
+                },
+                # "asl": {
+                #     "datatype": "perf",
+                #     "suffix": "asl",
+                #     "sidecars": ["asl.json"],
+                #     "task": "",
+                # },
+            }
+        ).apply(pd.Series)
+    )
+
```

Comment:
I think the second part could be changed like this:

```python
def compute_fieldmaps(df: pd.DataFrame) -> pd.DataFrame:
    filter = ["source_id", "source_ses_id", "modality", "dir_num", "suffix"]
    df1 = (
         df[filter][df["modality"].str.contains("fieldmap")].groupby(filter[:-1]).min()
     )
    df2 = (
         df[filter[:-1][df["modality"].str.contains("fieldmap")].groupby(filter[:-2]).min()
     )
    df1 = df1.join(df2.rename(columns={"dir_num": "run_01_dir_num"}))
    df_alt = df1.reset_index().assign(
         fmap_type_p=lambda x: (x.run_01_dir_num != x.dir_num)
     )
    return df_alt.assign(
         fmap=lambda x: x.fmap_type_p.apply(
             lambda y: "phasediff" if y else "magnitude"
         )
     )

def merge_fieldmaps(df, df_alt):
    filter = ["source_id", "source_ses_id", "modality", "dir_num", "fmap"]
    df1 = df.merge(df_alt[filter], how="inner", on=filter[:-1])
    df2 = df.merge(df_alt[filter], how="left", on=filter[:-1])
    df2 = df2[~df2["modality"].str.contains("fieldmap")]
    df1 = df1.assign(suffix=lambda x: x.fmap)
    return pd.concat([df2, df1], ignore_index=True)

def find_me_a_name(df: pd.DataFrame) -> pd.DataFrame:
    filter = ["source_id", "source_ses_id", "suffix", "dir_num"]
    df1 = df[filter].groupby(filter).min()
    df2 = df[filter].groupby(filter[:-1]).min()
    df1 = df1.join(df2.rename(columns={"dir_num": "run_01_dir_num"}))
    df_alt = df1.reset_index().assign(run=lambda x: (x.run_01_dir_num != x.dir_num))
    return df_alt.assign(
         run_num=lambda df: df.run.apply(lambda x: f"run-0{int(x)+1}")
     )
```

which gives:

```python
(...)
df_fmap = df.assign(
    dir_num=lambda x: x.source.apply(lambda y: int(Path(Path(y).parent).name))
 )
df_suf = merge_fieldmaps(df_fmap, compute_fieldmaps(df_fmap))
df_suf_dir = df_suf.assign(
     dir_num=lambda x: x.source.apply(lambda y: int(Path(Path(y).parent).name))
)
df_alt = find_me_a_name(df_suf_dir)
df_sub_ses_run = df_suf_dir.merge(
     df_alt[["source_id", "source_ses_id", "suffix", "dir_num", "run_num"]],
     how="left",
     on=["source_id", "source_ses_id", "suffix", "dir_num"],
)
(...)
```

Looking at the extracted functions, I think the code can be further simplified. We'll see if we can iterate further once the changes are incorporated and tested.